### PR TITLE
NEW: TPT3-88. Client Service Authorization

### PR DESCRIPTION
**Pull request by denys-davydov**

### CODE REVIEW

File: api-core-abac/src/main/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizer.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    final ClientCondition clientCondition = getClientCondition(client, perryAccount);
+    return authorizeClientReadOperation(clientCondition, staffPrivilegeTypes, client, perryAccount);
+  }
+
+  private ClientCondition getClientCondition(final Client client, final PerryAccount perryAccount) {
+    final Short clientCountyCode = countyDeterminationService.getClientCountyById(client.getIdentifier());
+    final Short staffGovernmentEntityType = getStaffGovernmentEntityType(perryAccount.getGovernmentEntityType());
+    return toClientCondition(client, clientCountyCode, staffGovernmentEntityType);
```
Comment (zemlyanoi) [inline api-core-abac/src/main/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizer.java]: What the purpose to add final in this case ? Method toClientCondition get final params
ClientCondition toClientCondition(final Client client, final Short clientCountyCode, final Short staffPersonCountyCode)
I think final in this case is redundant

File: api-core-abac/src/main/java/gov/ca/cwds/authorizer/util/StaffPrivilegeUtil.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  private static final String CWS_CASE_MANAGEMENT_SYSTEM = "CWS Case Management System";
+  private static final String STATE_OF_CALIFORNIA = "State of California";
+  private static final String SENSITIVE_PERSONS = "Sensitive Persons";
+  private static final String SEALED = "Sealed";
+
+  private StaffPrivilegeUtil() {
+  }
+
+  // TODO(dd): Consider to move this logic to drools
```
Comment (AlexanderSerbin) [inline api-core-abac/src/main/java/gov/ca/cwds/authorizer/util/StaffPrivilegeUtil.java]: This is not good practice to add TODO, You'd rather create JIRA task or implement it by yourself. Nobody but you would implement this. TODO will add Sonar violations, And agree similar Utils classes usually created when developer did not find proper place for the code

File: api-core-abac/src/main/java/gov/ca/cwds/authorizer/drools/DroolsAuthorizationService.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    final Collection<Object> facts = toFactsCollection(clientCondition, staffPrivilegeTypes);
+    return droolsService.performAuthorizationRules(
+        ClientAuthorizationDroolsConfiguration.INSTANCE,
+        facts
+    );
+  }
+
+  private Collection<Object> toFactsCollection(final ClientCondition clientCondition,
+      final Collection<StaffPrivilegeType> staffPrivilegeTypes) {
+    final Collection<Object> facts = new ArrayList<>(staffPrivilegeTypes.size() + 1);
```
Comment (zemlyanoi) [inline api-core-abac/src/main/java/gov/ca/cwds/authorizer/drools/DroolsAuthorizationService.java]: in this case you can use unmodifiableList if you want to use final

File: api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java [hunk 4]
```diff
@@ -1,179 +1,56 @@
...
-      session = getSessionFactory().getCurrentSession();
-    } catch (HibernateException e) { // NOSONAR
-      LOGGER.warn("NO SESSION!");
-      session = getSessionFactory().openSession();
-    }
+  private final SessionFactory sessionFactory;
 
-    return session;
+  @Inject
+  public CountyDeterminationDao(@CwsRsSessionFactory SessionFactory sessionFactory) {
```
Comment (AlexanderSerbin) [inline api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java]: Do you have reasons for utilizing RSCWS session factory instead CWS?

File: api-core-abac/src/test/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizerTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        droolsAuthorizationService,
+        clientCountyDeterminationServiceMock
+    );
+  }
+
+  @Test
+  public void checkInstance_returnFalse_whenPerryPrivilegesAreNull() {
+    final PerryAccount perryAccount = new PerryAccount();
+    final boolean expectedResult = false;
+    checkInstance_returnExpected_withPreparedPerryAccount(perryAccount, expectedResult);
```
Comment (AlexanderSerbin) [inline api-core-abac/src/test/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizerTest.java]: Method name is not compliant with Java conventions

File: drools-engine/src/main/java/gov/ca/cwds/drools/DroolsService.java [hunk 6]
```diff
@@ -1,35 +1,82 @@
...
 import org.kie.api.runtime.KieContainer;
 import org.kie.api.runtime.KieSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * @author CWDS CALS API Team
  */
 
 public class DroolsService {
```
Comment (AlexanderSerbin) [inline drools-engine/src/main/java/gov/ca/cwds/drools/DroolsService.java]: Consider method performBusinessRules and performAuthorizationRules to use the same code base

File: api-core-abac/src/main/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizer.java (hunk 1 above)
Comment (denys-davydov) [reply api-core-abac/src/main/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizer.java]: I use final with all the variables that are not planned to be reassigned.

File: api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java (hunk 4 above)
Comment (denys-davydov) [reply api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java]: Yes, this dao works with different than CWSCMS schema

File: api-core-abac/src/test/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizerTest.java (hunk 5 above)
Comment (denys-davydov) [reply api-core-abac/src/test/java/gov/ca/cwds/authorizer/ClientAbstractReadAuthorizerTest.java]: Such test methods' naming is considered to be the best practice in unit testing.

File: drools-engine/src/main/java/gov/ca/cwds/drools/DroolsService.java (hunk 6 above)
Comment (denys-davydov) [reply drools-engine/src/main/java/gov/ca/cwds/drools/DroolsService.java]: Let's discuss it offline

File: api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java [hunk 7]
```diff
@@ -1,179 +1,56 @@
...
-        + "FROM CASE_T A, CLN_RELT B\n"
-        + "WHERE (B.FKCLIENT_0 = :clientId\n"
-        + "       OR  B.FKCLIENT_T = :clientId)\n"
-        + "      AND (A.FKCHLD_CLT = B.FKCLIENT_T\n"
-        + "           OR  A.FKCHLD_CLT = B.FKCLIENT_0)\n"
-        + "      AND  A.END_DT IS NOT NULL ";
-
-    return executeNativeQueryAndReturnCountyList(namedQuery, clientId);
+  public List<Short> getClientCounties(final String clientId) {
+    final String nativeQuery = "SELECT GVR_ENTC "
```
Comment (zemlyanoi) [inline api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java]: This comment will not affect commit but

Use final where it reflects the fundamental semantics of your program: to indicate that classes are intended to be immutable or that fields are intended to be read-only.

Still I don't understand this dancing with final - no benefits - just an additional word

File: api-core-abac/src/main/java/gov/ca/cwds/authorizer/util/StaffPrivilegeUtil.java (hunk 2 above)
Comment (denys-davydov) [reply api-core-abac/src/main/java/gov/ca/cwds/authorizer/util/StaffPrivilegeUtil.java]: 1. The plan is to discuss this move with the team, TODO is a marker not to forget the issue. 
2. Renaming this class from StaffPrivilegeUtil to StaffPrivilegeMapper would be a solution I hope.

File: api-core-abac/src/main/java/gov/ca/cwds/authorizer/drools/DroolsAuthorizationService.java (hunk 3 above)
Comment (denys-davydov) [reply api-core-abac/src/main/java/gov/ca/cwds/authorizer/drools/DroolsAuthorizationService.java]: The list is modified in two next lines.

File: api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java (hunk 7 above)
Comment (denys-davydov) [reply api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java]: If we are talking about local variables only, the main benefit of using final is to highlight variables that are planned to be reassigned. While using final, reassignment of variables is restricted. This helps to read and support code better and faster.

Comment (zemlyanoi) [reply api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java]: It is not a benefit - it's just thinking about future changes
we don't have to think about future

Comment (denys-davydov) [reply api-core-abac/src/main/java/gov/ca/cwds/data/dao/cms/CountyDeterminationDao.java]: It is not about future changes. It shows right now if the variable will be reassigned or not.

**Pull request merged**